### DISCUSSION TITLE
SFTPGo as 3 replica set for achieving HA in GKE

### DISCUSSION THREAD

**Discussion by rashmimaharana** [drakkan/sftpgo#907] [Q&A]
Hi @drakkan 

We are trying to run SFTPGo as 3 replica set in GKE for achieving HA. It is backed by LB service. We are using gcs fuse mount to share the persistent data across all 3 pods which is mounted on "/srv/sftpgo/data/". 
We have installed gcsfuse  inside sftpgo docker image by modifying the official Dockerfile.

We were able to spin 3 pods and were able to transfer files. But its not consistent as out 10 files sometime 2 or 3 files are not getting transferred and errored out. I am posting the modified Dockerfile, Deployment.yaml and exception in this chat. Please let us know incase any configuration that we are missing.

Note : For sftp account we are setting UID and GID value which other than 0. 

Deployement.yaml file
------------------------
apiVersion: "apps/v1"
kind: "Deployment"
metadata:
  name: "sftpgo-server"
  namespace: "sftpgo"
  labels:
    app: "sftpgo-server"
spec:
  replicas: 3
  selector:
    matchLabels:
      app: "sftpgo-server"
  template:
    metadata:
      labels:
        app: "sftpgo-server"
    spec:
      containers:
      - name: "sftpgo-sha256-1"
        image: "us-east1-docker.pkg/sftpgo-fuse:2.2.3"
        imagePullPolicy: "Always"
        securityContext:
          privileged: true
          capabilities:
            add:
              - SYS_ADMIN
        env:
        - name: "SFTPGO_DATA_PROVIDER__DRIVER"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DATA_PROVIDER__DRIVER"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DATA_PROVIDER__NAME"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DATA_PROVIDER__NAME"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DATA_PROVIDER__HOST"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DATA_PROVIDER__HOST"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DATA_PROVIDER__PORT"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DATA_PROVIDER__PORT"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DATA_PROVIDER__USERNAME"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DATA_PROVIDER__USERNAME"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DATA_PROVIDER__PASSWORD"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DATA_PROVIDER__PASSWORD"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DATA_PROVIDER__CREATE_DEFAULT_ADMIN"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DATA_PROVIDER__CREATE_DEFAULT_ADMIN"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DEFAULT_ADMIN_USERNAME"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DEFAULT_ADMIN_USERNAME"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DEFAULT_ADMIN_PASSWORD"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DEFAULT_ADMIN_PASSWORD"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_HTTPD__SIGNING_PASSPHRASE"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_HTTPD__SIGNING_PASSPHRASE"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_SFTPD__BINDINGS__0__PORT"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_SFTPD__BINDINGS__0__PORT"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_HTTPD__BINDINGS__0__PORT"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_HTTPD__BINDINGS__0__PORT"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_TELEMETRY__BIND_PORT"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_TELEMETRY__BIND_PORT"
              name: "sftpgo-server-config-map"
        - name: "SFTPGO_DATA_PROVIDER__PREFER_DATABASE_CREDENTIALS"
          valueFrom:
            configMapKeyRef:
              key: "SFTPGO_DATA_PROVIDER__PREFER_DATABASE_CREDENTIALS"
              name: "sftpgo-server-config-map"
        - name: "GOOGLE_APPLICATION_CREDENTIALS"
          value: "/var/run/secret/cloud.google.com/service-account.json"
        lifecycle:
          postStart:
            exec:
              command:
              - gcsfuse
              - -o
              - nonempty
              - gcs-fuse
              - /srv/sftpgo  
        livenessProbe:
          failureThreshold: 3    
          httpGet:        
            path: /healthz       
            port: 8080    
          initialDelaySeconds: 15  
          periodSeconds: 15
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 8080
          initialDelaySeconds: 15
          periodSeconds: 15
        ports:
          - name: http-port
            containerPort: 8080
          - name: sftp-port
            containerPort: 2022 
        volumeMounts:
          - name: "service-account"
            mountPath: "/var/run/secret/cloud.google.com"
          - mountPath: /var/lib/sftpgo/id_ecdsa
            name: sftpgo-host-keys
            subPath: id_ecdsa
            readOnly: true
          - mountPath: /var/lib/sftpgo/id_ecdsa.pub
            name: sftpgo-host-keys
            subPath: id_ecdsa.pub
            readOnly: true
          - mountPath: /var/lib/sftpgo/id_ed25519
            name: sftpgo-host-keys
            subPath: id_ed25519
            readOnly: true
          - mountPath: /var/lib/sftpgo/id_ed25519.pub
            name: sftpgo-host-keys
            subPath: id_ed25519.pub
            readOnly: true
          - mountPath: /var/lib/sftpgo/id_rsa
            name: sftpgo-host-keys
            subPath: id_rsa
            readOnly: true
          - mountPath: /var/lib/sftpgo/id_rsa.pub
            name: sftpgo-host-keys
            subPath: id_rsa.pub
            readOnly: true
      volumes:
        - name: "service-account"
          secret:
            secretName: "sftpgo-sa-key"
        - name: sftpgo-host-keys
          secret:
            secretName: sftpgo-host-keys

Dockerfile
-----------
FROM golang:1.17-bullseye as builder

ENV GOFLAGS="-mod=readonly"

#RUN go get -u github.com/googlecloudplatform/gcsfuse

RUN mkdir -p /workspace
WORKDIR /workspace

ARG GOPROXY

COPY go.mod go.sum ./
RUN go mod download

ARG COMMIT_SHA

ARG FEATURES

COPY . .

RUN set -xe && \
    export COMMIT_SHA=${COMMIT_SHA:-$(git describe --always --dirty)} && \
    go build $(if [ -n "${FEATURES}" ]; then echo "-tags ${FEATURES}"; fi) -trimpath -ldflags "-s -w -X github.com/drakkan/sftpgo/v2/version.commit=${COMMIT_SHA} -X github.com/drakkan/sftpgo/v2/version.date=`date -u +%FT%TZ`" -v -o sftpgo

FROM debian:bullseye-slim

ARG INSTALL_OPTIONAL_PACKAGES=false

RUN apt-get update && apt-get install --no-install-recommends -y ca-certificates media-types && rm -rf /var/lib/apt/lists/*

RUN if [ "${INSTALL_OPTIONAL_PACKAGES}" = "true" ]; then apt-get update && apt-get install --no-install-recommends -y git && rm -rf /var/lib/apt/lists/*; fi

RUN apt-get update && apt-get install -y curl lsb-release sudo rsync gnupg
RUN export GCSFUSE_REPO=gcsfuse-jessie \
    && echo "deb http://packages.cloud.google.com/apt $GCSFUSE_REPO main" | tee /etc/apt/sources.list.d/gcsfuse.list \
    && curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
RUN apt-get update  && apt-get install -y gcsfuse

RUN mkdir -p /etc/sftpgo /var/lib/sftpgo /usr/share/sftpgo /srv/sftpgo/data /srv/sftpgo/backups

RUN groupadd --system -g 1000 sftpgo && \
    useradd --system --gid sftpgo --no-create-home \
    --home-dir /var/lib/sftpgo --shell /usr/sbin/nologin \
    --comment "SFTPGo user" --uid 1000 sftpgo

COPY --from=builder /workspace/sftpgo.json /etc/sftpgo/sftpgo.json
COPY --from=builder /workspace/templates /usr/share/sftpgo/templates
COPY --from=builder /workspace/static /usr/share/sftpgo/static
COPY --from=builder /workspace/openapi /usr/share/sftpgo/openapi
COPY --from=builder /workspace/sftpgo /usr/local/bin/
COPY --from=builder /go/bin/gcsfuse /usr/local/bin/

ENV SFTPGO_LOG_FILE_PATH=""
ENV SFTPGO_HTTPD__TEMPLATES_PATH=/usr/share/sftpgo/templates
ENV SFTPGO_SMTP__TEMPLATES_PATH=/usr/share/sftpgo/templates
ENV SFTPGO_HTTPD__STATIC_FILES_PATH=/usr/share/sftpgo/static
ENV SFTPGO_HTTPD__OPENAPI_PATH=/usr/share/sftpgo/openapi

RUN sed -i "s|\"users_base_dir\": \"\",|\"users_base_dir\": \"/srv/sftpgo/data\",|" /etc/sftpgo/sftpgo.json && \
    sed -i "s|\"backups\"|\"/srv/sftpgo/backups\"|" /etc/sftpgo/sftpgo.json

RUN chown -R sftpgo:sftpgo /etc/sftpgo /srv/sftpgo && chown sftpgo:sftpgo /var/lib/sftpgo && chmod 700 /srv/sftpgo/backups

WORKDIR /var/lib/sftpgo
USER 1000:1000

CMD ["sftpgo", "serve"]

Exception
----------
remote open("/usdal02-scp06.20220421150000.gz"): Failure
remote open("/v2_transfer.done"): Failure
remote open("/20220520/10/20220520110516/usdal02/usdal02-scp03.20220520100000.gz"): Failure
remote open("/20220520/10/20220520110516/usdal02/usdal02-scp04.20220520100000.gz"): No such file or directory
remote open("/20220520/12/20220520130614/usdal02/usdal02-gwy01.20220520120000.gz"): Failure

**Comment by drakkan** [comment]
Hi,

can you please explain why you are using gcsfuse? You can add a GCS bucket in SFTPGo directly, no fuse mount is required. 

In your Dockerfile you have:

```
USER 1000:1000
```

therefore SFTPGo cannot change uid/gid for files. It must be run as root user for this to work (this does not affect the problem you reported).

To understand the issue you have to look at your log files, but first I suggest to use SFTPGo in a proper way, without fuse mounts if they are not strictly necessary for reasons that you have not explained

**Reply by rashmimaharana** [reply]
Hi @drakkan 

Yes we tried without fuse mount. It worked fine for single POD. We are using GCS as storage location. But with fuse mount option we can use single bucket to share /srv/sftpgo/ directory across multiple POD.

Here is the problem we faced. We spin 2 POD and created one sftp user account. Upon user first login to sftp, lets say request goes to POD-1 and it creates home directory in /srv/sftpgo/data/USER-1. Then we tried to upload file, it gave us error "something like unable to perform the operation " as request re routed to another pod say POD-2. So we manually logged in to POD-2 shell and manually create the home directory for the USER-1 and then after that file transfer went fine. 
Then checked you official documentation and mounted the common data directory "/srv/sftpgo" across all PODs. In GKE we had to use fuse mount to mount GCS bucket inside container running in multiple POD as persistent volume.

Also can you please suggest the best practice and configuration required to run multiple sftpgo instance as container for HA. Will be really helpful.

https://github.com/drakkan/sftpgo/blob/main/docker/README.md

**Reply by drakkan** [reply]
Hi,

maybe I'm missing something obvious here. If you want to use a GCS bucket and 3 SFTPGo instance you can use:

- a shared database instance, for example a managed PostgreSQL or MySQL instance. This way all SFTPGo instances will share the same users
- you define SFTPGo users with the GCS bucket as backend. The bucket is already shared

You don't need shared data this way. You only need to share SFTP host keys that you can make available as docker secrets or something similar.

So my question is why is it necessary to share `/srv/sftpgo` on multiple pods? Do you need to share something other than user data? 

Many other SFTPGo users use the above architecture with no issues.
